File: src/main/ruby/g0101_0200/s0148_sort_list/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Sorting #Two_Pointers #Linked_List
# #Divide_and_Conquer #Merge_Sort #Level_2_Day_4_Linked_List #Big_O_Time_O(log(N))_Space_O(log(N))
# #2023_11_24_Time_188_ms_(84.21%)_Space_224_MB_(21.05%)

require_relative '../../com_github_leetcode/list_node'

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def sort_list(head)
  list = []
  new_head = ListNode.new

  while head
    list << head.val
    head = head.next
  end

  curr = new_head
  list.sort.each do |v|
    curr.next = ListNode.new(v)
    curr = curr.next
  end

  new_head.next
end

File: src/main/php/g0101_0200/s0148_sort_list/Solution.php
<?php

namespace leetcode\g0101_0200\s0148_sort_list;

// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Sorting #Two_Pointers #Linked_List
// #Divide_and_Conquer #Merge_Sort #Level_2_Day_4_Linked_List #Top_Interview_150_Divide_and_Conquer
// #Big_O_Time_O(log(N))_Space_O(log(N)) #2023_12_17_Time_63_ms_(97.30%)_Space_38.1_MB_(81.08%)

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {
    /**
     * @param ListNode $head
     * @return ListNode
     */
    public function sortList($head) {
        $sortArray = [];
        $sortedHead = $head;
        while (!is_null($sortedHead)) {
            $sortArray[] = $sortedHead->val;
            $sortedHead = $sortedHead->next;
        }
        unset($sortedHead);
        sort($sortArray);
        $newHead = $head;
        for ($i = 0; $i < count($sortArray); $i++) {
            $newHead->val = $sortArray[$i];
            $newHead = $newHead->next;
        }
        unset($newHead);
        return $head;
    }
}
